docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,150 @@
+    - The most generic rule seems to be the package's origin. With famous and well established public registries such as `maven.org` or `huggingface.co`, it is easy to correlate the package's origin with a more specific type.
+    - Mapping famous and public registries will prevent the use of private registries. To solve this, we need to allow users to change Cachi2's default configuration to enable additional source URLs for each purl type.
+    - More rules might be useful or even necessary when generating specific purl types. For example, Maven has the concept of groupId, artifactId and version (GAV) for each of its artifacts, and these values are fundamental in defining both the download URL and the purl.
+- Extend the generic artifacts lockfile specification to introduce a `type` attribute that allows users to hint at which purl type that artifact should have.
+    - The only types that can be selected are the ones that had their respective `enhancers` implemented. We don't want to allow users to be able to freely specify the purl type for a generic artifact, but rather restrict this use to a specific subset of purl types.
+    - Any failures to match the hinted `type` will not cause the request to fail. The file will still be fetched, but the purl will revert to `pkg:generic`.
+
+## A practical example
+
+### Input files
+
+**generic_artifacts.yaml**
+```yaml
+metadata:
+    version: '1.0'
+artifacts:
+  - download_url: https://github.com/containerbuildsystem/cachi2/archive/refs/tags/0.11.0.tar.gz
+    target: cachi2_0_11_0.tar.gz
+    checksums: 
+      sha256: fa0d536389db15fb3dabdb3b3d08354f47f765a653178140bfbe1b3de1a6ee76
+  - download_url: https://maven.repository.internal.com/ga/io/quarkus/quarkus-core/3.8.5.internal-00004/quarkus-core-3.8.5.internal-00004.jar
+    target: quakus.jar
+    type: maven
+    checksums:
+      sha1: e4ca5fadf89e62fb29d0d008046489b2305295bf
+  - download_url: https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/b919e5d07ce15f31ea741f2be99a00a33c3b427b/model-00001-of-00030.safetensors
+    target: llama_3.1_1_of_30.safetensors
+    type: huggingface
+```
+
+**.cachi2-config.yaml**
+```yaml
+generic-artifact-sources:
+    maven:
+        # sample internal registry
+        - maven.repository.internal.com
+        - maven.org
+    huggingface:
+        - huggingface.co
+```
+
+### Cachi2 CLI usage
+
+```
+cachi2 fetch-deps --source /path/to/repo generic
+```
+
+### Enhancer high-level definition
+
+- MavenPurlEnhancer:
+    - validates the origin URL
+    - parses the download url and converts it into the expected purl
+        ```bash
+        # sample url
+        https://maven.repository.internal.com/ga/io/quarkus/quarkus-core/3.8.5.internal-00004/quarkus-core-3.8.5.internal-00004.jar
+
+        # how the parsing will be done
+        https://{repository_url}/{as_dir(group_id)}/{artifactId}/{version}/{artifact_id}-{version}.{extension}
```

Comment:
This is missing a potential artifact `classifier`. With a `classifier` it would be
```
https://{repository_url}/{as_dir(group_id)}/{artifactId}/{version}/{artifact_id}-{version}[{-classifier}].{extension}
```
As mentioned in one of the comments above, if a `classifier` is present, you may struggle to figure out where the `classifier` ends and where the `extension` begins.

Reply:
I think we can mark this one as resolved as based on the https://github.com/containerbuildsystem/cachi2/pull/692#discussion_r1810861033 thread is seems like we'll need to keep adding more PURL-type related attributes.